File: Sources/TePlannerKit/Automations/Interpreters/RuleDisplay.swift
import Foundation

/// 把 declarative rule schema 里的 wire-format 字符串（entity 点路径、
/// trigger type、capability id 等）翻译成给用户看的中文。Source of truth
/// 暂时放 iOS 这边；后续如果矩阵变大就移到 backend 服务端 localization。
public enum RuleDisplay {
    /// 把观察项 entity 字符串翻译成用户友好的名字。
    public static func entityName(_ entity: String) -> String {
        switch entity {
        case "vehicle.climate.keeper_mode": return "空调保持模式"
        case "vehicle.sentry_mode_on": return "哨兵模式"
        case "vehicle.cabin_overheat_protection_on": return "座舱过热保护"
        case "vehicle.charging.state": return "充电状态"
        case "vehicle.battery_level": return "电量百分比"
        // Slice A — closure / lock virtual entities. Names embed
        // "停车后" because that's the precondition baked into the
        // virtual signal (filters out driving false-positives).
        case "vehicle.locked": return "车锁"
        case "vehicle.parked_unlocked": return "停车后未锁车"
        case "vehicle.parked_with_door_open": return "停车后车门开"
        case "vehicle.parked_with_window_open": return "停车后车窗开"
        case "vehicle.parked_with_frunk_open": return "停车后前备箱开"
        case "vehicle.parked_with_trunk_open": return "停车后后备箱开"
        default: return entity
        }
    }

    public static func triggerTypeName(_ type: String) -> String {
        switch type {
        case "state_duration": return "持续状态"
        case "state_transition": return "状态变化"
        case "cron": return "定时"
        case "geofence": return "地理围栏"
        default: return type
        }
    }

    /// 翻译 capability id 到操作名。空字符串 / "automation.dismiss" 视
    /// 为"仅关闭提醒"。
    public static func capabilityName(_ capabilityId: String) -> String {
        switch capabilityId {
        // Climate (climate.py + climate_extra.py)
        case "tesla.climate.set_keeper_mode":           return "调整空调保持模式"
        case "tesla.climate.preheat":                   return "启动预热"
        case "tesla.climate.stop":                      return "关闭空调"
        case "tesla.climate.set_temps":                 return "设置温度"
        case "tesla.climate.set_preconditioning_max":   return "切换最大预热"
        case "tesla.climate.set_cabin_overheat":        return "切换座舱过热保护"
        // Charging (charging.py + charging_extra.py)
        case "tesla.charging.set_limit":                return "调整充电限额"
        case "tesla.charging.start":                    return "开始充电"
        case "tesla.charging.stop":                     return "停止充电"
        case "tesla.charging.port_open":                return "打开充电口"
        case "tesla.charging.port_close":               return "关闭充电口"
        case "tesla.charging.set_amps":                 return "调整充电电流"
        // Security (security.py + closures.py)
        case "tesla.security.set_sentry":               return "切换哨兵模式"
        case "tesla.security.door_lock":                return "锁车"
        case "tesla.security.door_unlock":              return "解锁"
        case "tesla.security.actuate_frunk":            return "打开前备箱"
        case "tesla.security.actuate_trunk":            return "操作后备箱"
        // Closures (closures.py)
        case "tesla.closures.window_vent":              return "通风开窗"
        case "tesla.closures.window_close":             return "关闭车窗"
        case "tesla.closures.sun_roof_vent":            return "通风开天窗"
        case "tesla.closures.sun_roof_close":           return "关闭天窗"
        // Comfort (comfort.py)
        case "tesla.comfort.set_seat_heater":           return "设置座椅加热"
        case "tesla.comfort.set_steering_wheel_heater": return "切换方向盘加热"
        // Media (comfort.py + attention.py)
        case "tesla.media.toggle_playback":             return "切换车机播放"
        case "tesla.media.set_volume":                  return "设置车机音量"
        case "tesla.media.next_track":                  return "下一首"
        case "tesla.media.prev_track":                  return "上一首"
        // Navigation (navigation.py + attention.py)
        case "tesla.navigation.send":                   return "发送导航目的地"
        case "tesla.navigation.send_address":           return "发送地址到车"
        // Attention (attention.py)
        case "tesla.attention.flash_lights":            return "闪灯"
        case "tesla.attention.honk_horn":               return "鸣笛"
        case "tesla.attention.trigger_homelink":        return "触发 HomeLink"
        case "automation.dismiss", "":                  return "仅关闭提醒"
        default: return capabilityId
        }
    }

    /// Group capabilities into user-facing categories so the rule
    /// builder can render a sectioned picker. Order in the returned
    /// `categories` array drives display order; categories with no
    /// matching capability are filtered out by the caller.
    public static func capabilityCategory(_ capabilityId: String) -> CapabilityCategory {
        switch capabilityId {
        case let id where id.hasPrefix("tesla.climate."):    return .climate
        case let id where id.hasPrefix("tesla.charging."):   return .charging
        case let id where id.hasPrefix("tesla.security."),
             let id where id.hasPrefix("tesla.closures."):   return .security
        case let id where id.hasPrefix("tesla.comfort."):    return .comfort
        case let id where id.hasPrefix("tesla.media."):      return .media
        case let id where id.hasPrefix("tesla.navigation."): return .navigation
        case let id where id.hasPrefix("tesla.attention."):  return .attention
        default:                                              return .other
        }
    }

    public enum CapabilityCategory: String, CaseIterable {
        case climate, charging, security, comfort, media, navigation, attention, other

        public var label: String {
            switch self {
            case .climate:    return "空调与温度"
            case .charging:   return "充电"
            case .security:   return "安全与门窗"
            case .comfort:    return "座椅与方向盘"
            case .media:      return "车机媒体"
            case .navigation: return "导航"
            case .attention:  return "提示与车辆控制"
            case .other:      return "其他"
            }
        }

        /// SF Symbol name for the category — surfaces the section
        /// visually on the picker even when expanded.
        public var symbol: String {
            switch self {
            case .climate:    return "thermometer.medium"
            case .charging:   return "bolt.fill"
            case .security:   return "lock.shield.fill"
            case .comfort:    return "carseat.left.fill"
            case .media:      return "play.rectangle.fill"
            case .navigation: return "location.fill"
            case .attention:  return "light.beacon.max.fill"
            case .other:      return "square.grid.2x2"
            }
        }
    }

    /// 把 trigger 的 equals/to 值渲染成对应 entity 语境下用户看得懂的
    /// 形式。布尔 → 开/关，keeper_mode int → 露营/保持/宠物/关闭，等。
    public static func describeValue(_ value: JSONValue, entity: String) -> String {
        switch value {
        case .bool(let b):
            return b ? "开" : "关"
        case .int(let i):
            if entity == "vehicle.climate.keeper_mode" {
                switch i {
                case 0: return "关闭"
                case 1: return "保持"
                case 2: return "宠物模式"
                case 3: return "露营模式"
                default: return "\(i)"
                }
            }
            return "\(i)"
        case .string(let s): return s
        case .double(let d): return "\(d)"
        case .null: return "(空)"
        default: return "?"
        }
    }

    /// 把简单的 5-字段 cron expr 翻译成中文。覆盖 v1 builder 生成的
    /// 形态："分 时 * * 周几"。复杂表达式 fall-through 显示原 expr。
    public static func cronSentence(_ expr: String) -> String {
        let parts = expr.split(separator: " ").map(String.init)
        guard parts.count == 5 else { return "定时：\(expr)" }
        let (minute, hour, dom, month, weekday) = (parts[0], parts[1], parts[2], parts[3], parts[4])
        guard dom == "*", month == "*",
              let h = Int(hour), let m = Int(minute) else {
            return "定时：\(expr)"
        }
        let timeText = String(format: "%02d:%02d", h, m)
        let dayText: String
        if weekday == "*" {
            dayText = "每天"
        } else if weekday == "1-5" {
            dayText = "每个工作日"
        } else if weekday == "0,6" || weekday == "6,0" {
            dayText = "每个周末"
        } else {
            // Try comma-separated digit list "1,3,5" → 周一/周三/周五
            let cnDays = ["周日","周一","周二","周三","周四","周五","周六"]
            let nums = weekday.split(separator: ",").compactMap { Int($0) }
            if !nums.isEmpty, nums.allSatisfy({ (0...6).contains($0) }) {
                dayText = nums.map { cnDays[$0] }.joined(separator: "、")
            } else {
                dayText = "定时（\(weekday)）"
            }
        }
        return "\(dayText) \(timeText)"
    }

    /// Returns the next time a cron-triggered rule will fire, or nil
    /// for non-cron rules / unsupported expressions. Supports the v1
    /// 5-field shapes the iOS builder emits: 'M H * * W' where W ∈
    /// {*, '1-5', '0,6', comma-separated digits 0..6}. Walks forward
    /// up to 8 days from `referenceDate`. Asia/Shanghai time zone.
    public static func nextCronFire(
        spec: RuleSpec,
        referenceDate: Date = Date(),
        timeZone: TimeZone = TimeZone(identifier: "Asia/Shanghai") ?? .current
    ) -> Date? {
        guard let trigger = spec["trigger"]?.objectValue,
              trigger.string("type") == "cron",
              let expr = trigger.string("expr") else { return nil }
        let parts = expr.split(separator: " ").map(String.init)
        guard parts.count == 5,
              let minute = Int(parts[0]),
              let hour = Int(parts[1]),
              parts[2] == "*", parts[3] == "*" else { return nil }
        let allowedWeekdays: Set<Int>
        switch parts[4] {
        case "*":
            allowedWeekdays = Set(0...6)
        case "1-5":
            allowedWeekdays = [1, 2, 3, 4, 5]
        case "0,6", "6,0":
            allowedWeekdays = [0, 6]
        default:
            let nums = parts[4].split(separator: ",").compactMap { Int($0) }
            guard !nums.isEmpty, nums.allSatisfy({ (0...6).contains($0) }) else { return nil }
            allowedWeekdays = Set(nums)
        }

        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone
        for offset in 0...8 {
            guard let day = cal.date(byAdding: .day, value: offset, to: referenceDate) else { continue }
            let cronWeekday = (cal.component(.weekday, from: day) + 6) % 7
            guard allowedWeekdays.contains(cronWeekday) else { continue }
            var components = cal.dateComponents([.year, .month, .day], from: day)
            components.hour = hour
            components.minute = minute
            guard let target = cal.date(from: components) else { continue }
            if target > referenceDate { return target }
        }
        return nil
    }

    public static func formatDurationMinutes(_ minutes: Int) -> String {
        if minutes < 1 { return "不到 1 分钟" }
        if minutes < 60 { return "\(minutes) 分钟" }
        let h = minutes / 60
        let m = minutes % 60
        return m == 0 ? "\(h) 小时" : "\(h) 小时 \(m) 分钟"
    }

    /// Detail 视图预览动作正文用 —— 把模板里的 {duration_human} 等占
    /// 位符填上 representative 值（duration 用规则自己的阈值，
    /// battery_level 用 80% 作 sample）。运行时实际推送会填真实数值。
    public static func previewTemplate(_ template: String, spec: RuleSpec) -> String {
        var out = template
        if let trigger = spec["trigger"]?.objectValue {
            if let mins = trigger.int("for_minutes") {
                out = out.replacingOccurrences(of: "{duration_human}", with: formatDurationMinutes(mins))
                out = out.replacingOccurrences(of: "{duration_minutes}", with: "\(mins)")
            }
        }
        out = out.replacingOccurrences(of: "{battery_level}", with: "80")
        return out
    }

    /// Translate `state_transition` target values from their wire form
    /// (Tesla-side enum strings) to user-facing Chinese.
    public static func transitionTargetName(entity: String, target: String) -> String {
        switch entity {
        case "vehicle.charging.state":
            switch target {
            case "Complete":     return "充电完成"
            case "Charging":     return "充电中"
            case "Disconnected": return "未连接"
            case "Stopped":      return "已停止"
            case "Starting":     return "开始充电"
            case "NoPower":      return "无电源"
            default:             return target
            }
        case "vehicle.connectivity":
            switch target {
            case "CONNECTED":    return "在线"
            case "DISCONNECTED": return "离线"
            default:             return target
            }
        default:
            return target
        }
    }

    /// SF Symbol name for the rule's trigger type — used in list
    /// cards so each automation has the kind of glanceable visual
    /// vocabulary iOS 快捷指令 uses (different category = different
    /// icon and accent).
    public static func triggerSymbol(_ spec: RuleSpec) -> String {
        guard let trigger = spec["trigger"]?.objectValue,
              let type = trigger.string("type") else {
            return "bell.badge.fill"
        }
        switch type {
        case "cron":              return "clock.fill"
        case "geofence":          return "location.fill"
        case "state_transition":  return "arrow.right.arrow.left.circle.fill"
        case "state_duration":
            // Sub-classify by the entity to pick a meaningful glyph,
            // matching iOS 快捷指令's habit of changing icon by
            // category (time / location / device / battery / …).
            let entity = trigger.string("entity") ?? ""
            switch entity {
            case "vehicle.battery_level":               return "battery.25"
            case "vehicle.climate.keeper_mode":         return "moon.zzz.fill"
            case "vehicle.sentry_mode_on":              return "shield.lefthalf.filled"
            case "vehicle.cabin_overheat_protection_on": return "thermometer.sun.fill"
            case "vehicle.parked_unlocked":             return "lock.open.fill"
            case "vehicle.parked_with_door_open",
                 "vehicle.parked_with_window_open",
                 "vehicle.parked_with_frunk_open",
                 "vehicle.parked_with_trunk_open":      return "door.left.hand.open"
            case "vehicle.connectivity":                return "antenna.radiowaves.left.and.right"
            default:                                    return "bell.badge.fill"
            }
        default:                  return "bell.badge.fill"
        }
    }

    /// 把规则的"动作"翻译成一句简短描述，配合 triggerSentence 形成
    /// "当 X 持续 Y，那么 Z" 的快捷指令风格描述。
    public static func actionSentence(_ spec: RuleSpec) -> String {
        // Pick the action that would actually fire — for state_duration
        // rules that's actions_above[0] (the threshold-crossed alert);
        // for everything else it's actions[0]. Empty action list → "".
        let bucketKey: String
        if spec["actions"]?.arrayValue != nil {
            bucketKey = "actions"
        } else if spec["actions_above"]?.arrayValue != nil {
            bucketKey = "actions_above"
        } else if spec["actions_below"]?.arrayValue != nil {
            bucketKey = "actions_below"
        } else {
            return ""
        }
        guard let arr = spec[bucketKey]?.arrayValue, let first = arr.first?.objectValue else {
            return ""
        }
        let aType = first.string("type") ?? ""
        let title = first.string("title") ?? ""
        switch aType {
        case "notify":
            return title.isEmpty ? "推送通知" : "通知「\(title)」"
        case "notify_and_offer":
            let label = first.string("primary_action_label") ?? ""
            if title.isEmpty {
                return label.isEmpty ? "推送通知" : "通知 + \(label)"
            }
            return label.isEmpty ? "通知「\(title)」" : "通知「\(title)」+ \(label)"
        case "invoke":
            let cap = first.string("capability") ?? ""
            return capabilityName(cap)
        case "wait_for_state":
            let then = first["then"]?.objectValue
            let thenTitle = then?.string("title") ?? ""
            return thenTitle.isEmpty ? "条件满足后通知" : "条件满足后通知「\(thenTitle)」"
        default:
            return aType
        }
    }

    /// 触发条件一句话总结。Detail 视图 / Home 视图都用。
    public static func triggerSentence(_ spec: RuleSpec) -> String {
        guard let trigger = spec["trigger"]?.objectValue,
              let type = trigger.string("type") else {
            return ""
        }
        switch type {
        case "state_duration":
            let entityRaw = trigger.string("entity") ?? ""
            let entity = entityName(entityRaw)
            let mins = trigger.int("for_minutes") ?? 0
            let op = trigger.string("op") ?? "=="
            switch op {
            case "<":
                let v = trigger["value"]?.intValue ?? 0
                return "「\(entity)」低于 \(v)% 持续 \(formatDurationMinutes(mins))"
            case "<=":
                let v = trigger["value"]?.intValue ?? 0
                return "「\(entity)」不超过 \(v)% 持续 \(formatDurationMinutes(mins))"
            case ">":
                let v = trigger["value"]?.intValue ?? 0
                return "「\(entity)」高于 \(v)% 持续 \(formatDurationMinutes(mins))"
            case ">=":
                let v = trigger["value"]?.intValue ?? 0
                return "「\(entity)」不低于 \(v)% 持续 \(formatDurationMinutes(mins))"
            default:
                // For boolean-true 虚拟实体（停车后未锁车 / 哨兵模式 等）
                // a "处于「开」状态" framing reads awkwardly. Prefer a
                // shorter form when the entity name itself describes
                // the active condition.
                let equals = trigger["equals"]
                if case .bool(let b) = equals ?? .null, b {
                    return "「\(entity)」持续 \(formatDurationMinutes(mins))"
                }
                let value = equals.flatMap { describeValue($0, entity: entityRaw) } ?? "?"
                return "「\(entity)」=「\(value)」 持续 \(formatDurationMinutes(mins))"
            }
        case "state_transition":
            let entity = entityName(trigger.string("entity") ?? "")
            let target = trigger.string("to") ?? "?"
            return "「\(entity)」变为「\(transitionTargetName(entity: trigger.string("entity") ?? "", target: target))」"
        case "cron":
            return cronSentence(trigger.string("expr") ?? "")
        case "geofence":
            let event = trigger.string("event") ?? "enter"
            let radius = trigger.int("radius_m") ?? 0
            let verb = event == "enter" ? "进入" : "离开"
            let lat = trigger.double("lat") ?? 0
            let lng = trigger.double("lng") ?? 0
            // 0,0 is the placeholder for unconfigured presets — call
            // out the user-action needed instead of leaking coords.
            if abs(lat) < 0.0001 && abs(lng) < 0.0001 {
                return "\(verb)某个区域 · \(radius)m（待选择地点）"
            }
            return "\(verb)区域 \(radius)m 范围"
        default:
            return type
        }
    }
}

File: apps/android/app/src/main/java/cloud/teplanner/android/automations/RuleDisplay.kt
package cloud.teplanner.android.automations

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Port of iOS RuleDisplay (capability name + category subset).
 * Source-of-truth for the user-facing string of each Tesla capability
 * id and the section it belongs to in the picker.
 *
 * Sync with iOS `Sources/TePlannerKit/Automations/Interpreters/RuleDisplay.swift`
 * + the backend `services/capabilities/tesla/` registry whenever a
 * new capability is added.
 */
object RuleDisplay {

    fun capabilityName(id: String): String = when (id) {
        // Climate
        "tesla.climate.set_keeper_mode" -> "调整空调保持模式"
        "tesla.climate.preheat" -> "启动预热"
        "tesla.climate.stop" -> "关闭空调"
        "tesla.climate.set_temps" -> "设置温度"
        "tesla.climate.set_preconditioning_max" -> "切换最大预热"
        "tesla.climate.set_cabin_overheat" -> "切换座舱过热保护"
        // Charging
        "tesla.charging.set_limit" -> "调整充电限额"
        "tesla.charging.start" -> "开始充电"
        "tesla.charging.stop" -> "停止充电"
        "tesla.charging.port_open" -> "打开充电口"
        "tesla.charging.port_close" -> "关闭充电口"
        "tesla.charging.set_amps" -> "调整充电电流"
        // Security
        "tesla.security.set_sentry" -> "切换哨兵模式"
        "tesla.security.door_lock" -> "锁车"
        "tesla.security.door_unlock" -> "解锁"
        "tesla.security.actuate_frunk" -> "打开前备箱"
        "tesla.security.actuate_trunk" -> "操作后备箱"
        // Closures
        "tesla.closures.window_vent" -> "通风开窗"
        "tesla.closures.window_close" -> "关闭车窗"
        "tesla.closures.sun_roof_vent" -> "通风开天窗"
        "tesla.closures.sun_roof_close" -> "关闭天窗"
        // Comfort
        "tesla.comfort.set_seat_heater" -> "设置座椅加热"
        "tesla.comfort.set_steering_wheel_heater" -> "切换方向盘加热"
        // Media
        "tesla.media.toggle_playback" -> "切换车机播放"
        "tesla.media.set_volume" -> "设置车机音量"
        "tesla.media.next_track" -> "下一首"
        "tesla.media.prev_track" -> "上一首"
        // Navigation
        "tesla.navigation.send" -> "发送导航目的地"
        "tesla.navigation.send_address" -> "发送地址到车"
        // Attention
        "tesla.attention.flash_lights" -> "闪灯"
        "tesla.attention.honk_horn" -> "鸣笛"
        "tesla.attention.trigger_homelink" -> "触发 HomeLink"
        "automation.dismiss", "" -> "仅关闭提醒"
        else -> id
    }

    enum class CapabilityCategory(val label: String) {
        CLIMATE("空调与温度"),
        CHARGING("充电"),
        SECURITY("安全与门窗"),
        COMFORT("座椅与方向盘"),
        MEDIA("车机媒体"),
        NAVIGATION("导航"),
        ATTENTION("提示与车辆控制"),
        OTHER("其他"),
    }

    fun capabilityCategory(id: String): CapabilityCategory = when {
        id.startsWith("tesla.climate.") -> CapabilityCategory.CLIMATE
        id.startsWith("tesla.charging.") -> CapabilityCategory.CHARGING
        id.startsWith("tesla.security.") || id.startsWith("tesla.closures.") ->
            CapabilityCategory.SECURITY
        id.startsWith("tesla.comfort.") -> CapabilityCategory.COMFORT
        id.startsWith("tesla.media.") -> CapabilityCategory.MEDIA
        id.startsWith("tesla.navigation.") -> CapabilityCategory.NAVIGATION
        id.startsWith("tesla.attention.") -> CapabilityCategory.ATTENTION
        else -> CapabilityCategory.OTHER
    }
}


/**
 * Static lookup tables — direct port of iOS CapabilityDefaults.
 * When a new capability is added to the backend registry, add a row
 * here too so the picker auto-populates sane defaults + chooses a
 * short button label.
 */
object CapabilityDefaults {

    /** Default `params` dict that auto-seeds when the user picks a
     *  capability for notify_and_offer. */
    val params: Map<String, JsonObject> = mapOf(
        "tesla.climate.set_keeper_mode" to JsonObject(mapOf("mode" to JsonPrimitive(0))),
        "tesla.climate.set_temps" to JsonObject(mapOf(
            "driver_temp" to JsonPrimitive(22.0),
            "passenger_temp" to JsonPrimitive(22.0),
        )),
        "tesla.climate.set_preconditioning_max" to JsonObject(mapOf("on" to JsonPrimitive(true))),
        "tesla.climate.set_cabin_overheat" to JsonObject(mapOf("mode" to JsonPrimitive(2))),
        "tesla.charging.set_limit" to JsonObject(mapOf("percent" to JsonPrimitive(80))),
        "tesla.charging.set_amps" to JsonObject(mapOf("amps" to JsonPrimitive(16))),
        "tesla.security.set_sentry" to JsonObject(mapOf("on" to JsonPrimitive(false))),
        "tesla.comfort.set_seat_heater" to JsonObject(mapOf(
            "seat" to JsonPrimitive(0), "level" to JsonPrimitive(2),
        )),
        "tesla.comfort.set_steering_wheel_heater" to JsonObject(mapOf("on" to JsonPrimitive(true))),
        "tesla.media.set_volume" to JsonObject(mapOf("volume" to JsonPrimitive(5.0))),
    )

    /** Short verb (≤4 Chinese chars) for the inline notification
     *  action button — distinct from `RuleDisplay.capabilityName`
     *  which is descriptive. */
    val buttonLabel: Map<String, String> = mapOf(
        "tesla.climate.set_keeper_mode" to "关闭模式",
        "tesla.climate.preheat" to "预热",
        "tesla.climate.stop" to "关空调",
        "tesla.climate.set_temps" to "调温",
        "tesla.climate.set_preconditioning_max" to "最大预热",
        "tesla.climate.set_cabin_overheat" to "切换",
        "tesla.charging.set_limit" to "调限额",
        "tesla.charging.start" to "开始充电",
        "tesla.charging.stop" to "停止充电",
        "tesla.charging.port_open" to "开充电口",
        "tesla.charging.port_close" to "关充电口",
        "tesla.charging.set_amps" to "调电流",
        "tesla.security.set_sentry" to "切换哨兵",
        "tesla.security.door_lock" to "锁车",
        "tesla.security.door_unlock" to "解锁",
        "tesla.security.actuate_frunk" to "前备箱",
        "tesla.security.actuate_trunk" to "后备箱",
        "tesla.closures.window_vent" to "通风",
        "tesla.closures.window_close" to "关车窗",
        "tesla.closures.sun_roof_vent" to "天窗通风",
        "tesla.closures.sun_roof_close" to "关天窗",
        "tesla.comfort.set_seat_heater" to "座椅加热",
        "tesla.comfort.set_steering_wheel_heater" to "方向盘加热",
        "tesla.media.toggle_playback" to "播/暂停",
        "tesla.media.set_volume" to "调音量",
        "tesla.media.next_track" to "下一首",
        "tesla.media.prev_track" to "上一首",
        "tesla.navigation.send" to "导航",
        "tesla.navigation.send_address" to "导航",
        "tesla.attention.flash_lights" to "闪灯",
        "tesla.attention.honk_horn" to "鸣笛",
        "tesla.attention.trigger_homelink" to "HomeLink",
    )
}
